src/server/pkg/storage/chunk/client.go

```diff
@@ -0,0 +1,173 @@
+package chunk
+
+import (
+	"context"
+	fmt "fmt"
+	io "io"
+	"io/ioutil"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/tracker"
+)
+
+// Client allows manipulation of individual chunks, by maintaining consistency between
+// a tracker and an obj.Client.
+type Client struct {
+	objc     obj.Client
+	mdstore  MetadataStore
+	tracker  tracker.Tracker
+	chunkSet string
+	ttl      time.Duration
+
+	mu sync.Mutex
+	n  int
+
+	cancel context.CancelFunc
+	err    error
+	done   chan struct{}
+}
+
+func NewClient(objc obj.Client, mdstore MetadataStore, tracker tracker.Tracker, chunkSet string) *Client {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	c := &Client{
+		objc:     objc,
+		tracker:  tracker,
+		mdstore:  mdstore,
+		chunkSet: chunkSet,
+		ttl:      defaultChunkTTL,
+		done:     make(chan struct{}),
+		cancel:   cancel,
+	}
+	if chunkSet != "" {
+		go func() {
+			c.err = c.runLoop(ctx)
+			close(c.done)
+		}()
+	} else {
+		close(c.done)
+	}
+	return c
+}
+
+func (c *Client) Create(ctx context.Context, md ChunkMetadata, r io.Reader) (ChunkID, error) {
+	chunkData, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	chunkID := Hash(chunkData)
+	var pointsTo []string
+	for _, cid := range md.PointsTo {
+		pointsTo = append(pointsTo, ChunkObjectID(cid))
+	}
+	// TODO: retry on ErrTombstone
+	chunkOID := ChunkObjectID(chunkID)
+	if err := c.tracker.CreateObject(ctx, chunkOID, pointsTo, c.ttl); err != nil {
+		if err != tracker.ErrObjectExists {
+			return nil, err
+		}
+	}
+	// create an object whos sole purpose is to reference the chunk we created, and to have a structured name
+	// which can be renewed in bulk by prefix
+	n := c.getInt()
+	if err := c.tracker.CreateObject(ctx, fmt.Sprintf("client-tmp/%s/%d", c.chunkSet, n), []string{chunkOID}, c.ttl); err != nil {
+		return nil, err
+	}
+	// at this point no one will be trying to delete the chunk, because there is an object pointing to it.
+	p := chunkPath(chunkID)
+	if c.objc.Exists(ctx, p) {
+		return chunkID, nil
+	}
+	if err := c.mdstore.SetChunkMetadata(ctx, chunkID, md); err != nil {
+		return nil, err
+	}
+	objW, err := c.objc.Writer(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+	defer objW.Close()
```

Comment:
We should capture any errors that occur here. Also, why are we closing twice?

Reply:
Yeah I guess we don't need to.